File: bindings/java/src/main/com/apple/foundationdb/TenantManagement.java
/*
 * TenantManagement.java
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.apple.foundationdb;

import java.nio.charset.Charset;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.Executor;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.BiFunction;

import com.apple.foundationdb.async.AsyncIterable;
import com.apple.foundationdb.async.AsyncIterator;
import com.apple.foundationdb.async.AsyncUtil;
import com.apple.foundationdb.async.CloseableAsyncIterator;
import com.apple.foundationdb.tuple.ByteArrayUtil;
import com.apple.foundationdb.tuple.Tuple;

/**
 * The FoundationDB API includes function to manage the set of tenants in a cluster.
 */
public class TenantManagement {
	static byte[] TENANT_MAP_PREFIX = ByteArrayUtil.join(new byte[] { (byte)255, (byte)255 },
	                                                     "/management/tenant/map/".getBytes());

	/**
	 * Creates a new tenant in the cluster. If the tenant already exists, this operation will complete
	 * successfully without changing anything. The transaction must be committed for the creation to take
	 * effect or to observe any errors.
	 *
	 * @param tr The transaction used to create the tenant.
	 * @param tenantName The name of the tenant. Can be any byte string that does not begin a 0xFF byte.
	 */
	public static void createTenant(Transaction tr, byte[] tenantName) {
		tr.options().setSpecialKeySpaceEnableWrites();
		tr.set(ByteArrayUtil.join(TENANT_MAP_PREFIX, tenantName), new byte[0]);
	}

	/**
	 * Creates a new tenant in the cluster. If the tenant already exists, this operation will complete
	 * successfully without changing anything. The transaction must be committed for the creation to take
	 * effect or to observe any errors.<br>
	 * <br>
	 * This is a convenience method that generates the tenant name by packing a {@code Tuple}.
	 *
	 * @param tr The transaction used to create the tenant.
	 * @param tenantName The name of the tenant, as a Tuple.
	 */
	public static void createTenant(Transaction tr, Tuple tenantName) {
		createTenant(tr, tenantName.pack());
	}

	/**
	 * Creates a new tenant in the cluster using a transaction created on the specified {@code Database}.
	 * This operation will first check whether the tenant exists, and if it does it will set the
	 * {@code CompletableFuture} to a tenant_already_exists error. Otherwise, it will attempt to create
	 * the tenant in a retry loop. If the tenant is created concurrently by another transaction, this
	 * function may still return successfully.
	 *
	 * @param db The database used to create a transaction for creating the tenant.
	 * @param tenantName The name of the tenant. Can be any byte string that does not begin a 0xFF byte.
	 * @return a {@code CompletableFuture} that when set without error will indicate that the tenant has
	 * been created.
	 */
	public static CompletableFuture<Void> createTenant(Database db, byte[] tenantName) {
		final AtomicBoolean checkedExistence = new AtomicBoolean(false);
		final byte[] key = ByteArrayUtil.join(TENANT_MAP_PREFIX, tenantName);
		return db.runAsync(tr -> {
			tr.options().setSpecialKeySpaceEnableWrites();
			if(checkedExistence.get()) {
				tr.set(key, new byte[0]);
				return CompletableFuture.completedFuture(null);
			}
			else {
				return tr.get(key).thenAcceptAsync(result -> {
					checkedExistence.set(true);
					if(result != null) {
						throw new FDBException("A tenant with the given name already exists", 2132);
					}
					tr.set(key, new byte[0]);
				});
			}
		});
	}

	/**
	 * Creates a new tenant in the cluster using a transaction created on the specified {@code Database}.
	 * This operation will first check whether the tenant exists, and if it does it will set the
	 * {@code CompletableFuture} to a tenant_already_exists error. Otherwise, it will attempt to create
	 * the tenant in a retry loop. If the tenant is created concurrently by another transaction, this
	 * function may still return successfully.<br>
	 * <br>
	 * This is a convenience method that generates the tenant name by packing a {@code Tuple}.
	 *
	 * @param db The database used to create a transaction for creating the tenant.
	 * @param tenantName The name of the tenant, as a Tuple.
	 * @return a {@code CompletableFuture} that when set without error will indicate that the tenant has
	 * been created.
	 */
	public static CompletableFuture<Void> createTenant(Database db, Tuple tenantName) {
		return createTenant(db, tenantName.pack());
	}

	/**
	 * Deletes a tenant from the cluster. If the tenant does not exists, this operation will complete
	 * successfully without changing anything. The transaction must be committed for the deletion to take
	 * effect or to observe any errors.<br>
	 * <br>
	 * <b>Note:</b> A tenant cannot be deleted if it has any data in it. To delete a non-empty tenant, you must
	 * first use a clear operation to delete all of its keys.
	 *
	 * @param tr The transaction used to delete the tenant.
	 * @param tenantName The name of the tenant being deleted.
	 */
	public static void deleteTenant(Transaction tr, byte[] tenantName) {
		tr.options().setSpecialKeySpaceEnableWrites();
		tr.clear(ByteArrayUtil.join(TENANT_MAP_PREFIX, tenantName));
	}

	/**
	 * Deletes a tenant from the cluster. If the tenant does not exists, this operation will complete
	 * successfully without changing anything. The transaction must be committed for the deletion to take
	 * effect or to observe any errors.<br>
	 * <br>
	 * <b>Note:</b> A tenant cannot be deleted if it has any data in it. To delete a non-empty tenant, you must
	 * first use a clear operation to delete all of its keys.<br>
	 * <br>
	 * This is a convenience method that generates the tenant name by packing a {@code Tuple}.
	 *
	 * @param tr The transaction used to delete the tenant.
	 * @param tenantName The name of the tenant being deleted, as a Tuple.
	 */
	public static void deleteTenant(Transaction tr, Tuple tenantName) {
		deleteTenant(tr, tenantName.pack());
	}

	/**
	 * Deletes a tenant from the cluster using a transaction created on the specified {@code Database}. This
	 * operation will first check whether the tenant exists, and if it does not it will set the
	 * {@code CompletableFuture} to a tenant_not_found error. Otherwise, it will attempt to delete the
	 * tenant in a retry loop. If the tenant is deleted concurrently by another transaction, this function may
	 * still return successfully.<br>
	 * <br>
	 * <b>Note:</b> A tenant cannot be deleted if it has any data in it. To delete a non-empty tenant, you must
	 * first use a clear operation to delete all of its keys.
	 *
	 * @param db The database used to create a transaction for deleting the tenant.
	 * @param tenantName The name of the tenant being deleted.
	 * @return a {@code CompletableFuture} that when set without error will indicate that the tenant has
	 * been deleted.
	 */
	public static CompletableFuture<Void> deleteTenant(Database db, byte[] tenantName) {
		final AtomicBoolean checkedExistence = new AtomicBoolean(false);
		final byte[] key = ByteArrayUtil.join(TENANT_MAP_PREFIX, tenantName);
		return db.runAsync(tr -> {
			tr.options().setSpecialKeySpaceEnableWrites();
			if(checkedExistence.get()) {
				tr.clear(key);
				return CompletableFuture.completedFuture(null);
			}
			else {
				return tr.get(key).thenAcceptAsync(result -> {
					checkedExistence.set(true);
					if(result == null) {
						throw new FDBException("Tenant does not exist", 2131);
					}
					tr.clear(key);
				});
			}
		});
	}

	/**
	 * Deletes a tenant from the cluster using a transaction created on the specified {@code Database}. This
	 * operation will first check whether the tenant exists, and if it does not it will set the
	 * {@code CompletableFuture} to a tenant_not_found error. Otherwise, it will attempt to delete the
	 * tenant in a retry loop. If the tenant is deleted concurrently by another transaction, this function may
	 * still return successfully.<br>
	 * <br>
	 * <b>Note:</b> A tenant cannot be deleted if it has any data in it. To delete a non-empty tenant, you must
	 * first use a clear operation to delete all of its keys.<br>
	 * <br>
	 * This is a convenience method that generates the tenant name by packing a {@code Tuple}.
	 *
	 * @param db The database used to create a transaction for deleting the tenant.
	 * @param tenantName The name of the tenant being deleted.
	 * @return a {@code CompletableFuture} that when set without error will indicate that the tenant has
	 * been deleted.
	 */
	public static CompletableFuture<Void> deleteTenant(Database db, Tuple tenantName) {
		return deleteTenant(db, tenantName.pack());
	}


	/**
	 * Lists all tenants in between the range specified. The number of tenants listed can be restricted.
	 *
	 * @param db The database used to create a transaction for listing the tenants.
	 * @param begin The beginning of the range of tenants to list.
	 * @param end The end of the range of the tenants to list.
	 * @param limit The maximum number of tenants to return from this request.
	 * @return an iterator where each item is a KeyValue object where the key is the tenant name
	 * and the value is the unprocessed JSON string containing the tenant's metadata
	 */
	public static CloseableAsyncIterator<KeyValue> listTenants(Database db, byte[] begin, byte[] end, int limit) {
		return listTenants_internal(db.createTransaction(), begin, end, limit);
	}

	/**
	 * Lists all tenants in between the range specified. The number of tenants listed can be restricted.
	 * This is a convenience method that generates the begin and end ranges by packing two {@code Tuple}s.
	 *
	 * @param db The database used to create a transaction for listing the tenants.
	 * @param begin The beginning of the range of tenants to list.
	 * @param end The end of the range of the tenants to list.
	 * @param limit The maximum number of tenants to return from this request.
	 * @return an iterator where each item is a KeyValue object where the key is the tenant name
	 * and the value is the unprocessed JSON string containing the tenant's metadata
	 */
	public static CloseableAsyncIterator<KeyValue> listTenants(Database db, Tuple begin, Tuple end, int limit) {
		return listTenants_internal(db.createTransaction(), begin.pack(), end.pack(), limit);
	}

	private static CloseableAsyncIterator<KeyValue> listTenants_internal(Transaction tr, byte[] begin, byte[] end,
	                                                                   int limit) {
		return new TenantAsyncIterator(tr, begin, end, limit);
	}

	// Templates taken from BoundaryIterator LocalityUtil.java
	static class TenantAsyncIterator implements CloseableAsyncIterator<KeyValue> {
		Transaction tr;
		final byte[] begin;
		final byte[] end;

		final AsyncIterable<KeyValue> firstGet;
		AsyncIterator<KeyValue> iter;
		private boolean closed;

		TenantAsyncIterator(Transaction tr, byte[] begin, byte[] end, int limit) {
			this.tr = tr;

			this.begin = ByteArrayUtil.join(TENANT_MAP_PREFIX, begin);
			this.end = ByteArrayUtil.join(TENANT_MAP_PREFIX, end);

			tr.options().setRawAccess();
			tr.options().setLockAware();

			firstGet = tr.getRange(this.begin, this.end, limit);
			iter = firstGet.iterator();
			closed = false;
		}

		@Override
		public CompletableFuture<Boolean> onHasNext() {
			return iter.onHasNext();
		}

		@Override
		public boolean hasNext() {
			return iter.hasNext();
		}
		@Override
		public KeyValue next() {
			KeyValue kv = iter.next();
			byte[] tenant = Arrays.copyOfRange(kv.getKey(), TENANT_MAP_PREFIX.length, kv.getKey().length);
			byte[] value = kv.getValue();

			KeyValue result = new KeyValue(tenant, value);
			return result;
		}

		@Override
		public void remove() {
			throw new UnsupportedOperationException("Tenant lists are read-only");
		}

		@Override
		public void close() {
			TenantAsyncIterator.this.tr.close();
			closed = true;
		}

		@Override
		protected void finalize() throws Throwable {
			try {
				if (FDB.instance().warnOnUnclosed && !closed) {
					System.err.println("CloseableAsyncIterator not closed (listTenants)");
				}
				if (!closed) {
					close();
				}
			} finally {
				super.finalize();
			}
		}
	}

	private TenantManagement() {}
}

File: bindings/python/fdb/tenant_management.py
#
# tenant_management.py
#
# This source file is part of the FoundationDB open source project
#
# Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# FoundationDB Python API

"""Documentation for this API can be found at
https://apple.github.io/foundationdb/api-python.html"""

from fdb import impl as _impl

_tenant_map_prefix = b"\xff\xff/management/tenant/map/"


# If the existence_check_marker is an empty list, then check whether the tenant exists.
# After the check, append an item to the existence_check_marker list so that subsequent
# calls to this function will not perform the existence check.
#
# If the existence_check_marker is a non-empty list, return None.
def _check_tenant_existence(tr, key, existence_check_marker, force_maybe_commited):
    if not existence_check_marker:
        existing_tenant = tr[key].wait()
        existence_check_marker.append(None)
        if force_maybe_commited:
            raise _impl.FDBError(1021)  # maybe_committed
        return existing_tenant != None

    return None


# Attempt to create a tenant in the cluster. If existence_check_marker is an empty
# list, then this function will check if the tenant already exists and fail if it does.
# Once the existence check is completed, it will not be done again if this function
# retries. As a result, this function may return successfully if the tenant is created
# by someone else concurrently. This behavior allows the operation to be idempotent with
# respect to retries.
#
# If the existence_check_marker is a non-empty list, then the existence check is skipped.
@_impl.transactional
def _create_tenant_impl(
    tr, tenant_name, existence_check_marker, force_existence_check_maybe_committed=False
):
    tr.options.set_special_key_space_enable_writes()
    key = b"%s%s" % (_tenant_map_prefix, tenant_name)

    if (
        _check_tenant_existence(
            tr, key, existence_check_marker, force_existence_check_maybe_committed
        )
        is True
    ):
        raise _impl.FDBError(2132)  # tenant_already_exists

    tr[key] = b""


# Attempt to delete a tenant from the cluster. If existence_check_marker is an empty
# list, then this function will check if the tenant already exists and fail if it does
# not. Once the existence check is completed, it will not be done again if this function
# retries. As a result, this function may return successfully if the tenant is deleted
# by someone else concurrently. This behavior allows the operation to be idempotent with
# respect to retries.
#
# If the existence_check_marker is a non-empty list, then the existence check is skipped.
@_impl.transactional
def _delete_tenant_impl(
    tr, tenant_name, existence_check_marker, force_existence_check_maybe_committed=False
):
    tr.options.set_special_key_space_enable_writes()
    key = b"%s%s" % (_tenant_map_prefix, tenant_name)

    if (
        _check_tenant_existence(
            tr, key, existence_check_marker, force_existence_check_maybe_committed
        )
        is False
    ):
        raise _impl.FDBError(2131)  # tenant_not_found

    del tr[key]


class FDBTenantList(object):
    """Iterates over the results of list_tenants query. Returns
    KeyValue objects.

    """

    def __init__(self, rangeresult):
        self._range = rangeresult
        self._iter = iter(self._range)

    def to_list(self):
        return list(self.__iter__())

    def __iter__(self):
        for next_item in self._iter:
            tenant_name = _impl.remove_prefix(next_item.key, _tenant_map_prefix)
            yield _impl.KeyValue(tenant_name, next_item.value)


# Lists the tenants created in the cluster, specified by the begin and end range.
# Also limited in number of results by the limit parameter.
# Returns an iterable object that yields KeyValue objects
# where the keys are the tenant names and the values are the unprocessed
# JSON strings of the tenant metadata
@_impl.transactional
def _list_tenants_impl(tr, begin, end, limit):
    tr.options.set_raw_access()
    begin_key = b"%s%s" % (_tenant_map_prefix, begin)
    end_key = b"%s%s" % (_tenant_map_prefix, end)

    rangeresult = tr.get_range(begin_key, end_key, limit)

    return FDBTenantList(rangeresult)


def create_tenant(db_or_tr, tenant_name):
    tenant_name = _impl.process_tenant_name(tenant_name)

    # Only perform the existence check when run using a database
    # Callers using a transaction are expected to check existence themselves if required
    existence_check_marker = (
        [] if not isinstance(db_or_tr, _impl.TransactionRead) else [None]
    )
    _create_tenant_impl(db_or_tr, tenant_name, existence_check_marker)


def delete_tenant(db_or_tr, tenant_name):
    tenant_name = _impl.process_tenant_name(tenant_name)

    # Only perform the existence check when run using a database
    # Callers using a transaction are expected to check existence themselves if required
    existence_check_marker = (
        [] if not isinstance(db_or_tr, _impl.TransactionRead) else [None]
    )
    _delete_tenant_impl(db_or_tr, tenant_name, existence_check_marker)


def list_tenants(db_or_tr, begin, end, limit):
    begin = _impl.process_tenant_name(begin)
    end = _impl.process_tenant_name(end)

    return _list_tenants_impl(db_or_tr, begin, end, limit)
